File: shared/utils/error_handlers.ts
export class ApplicationError extends Error {
  public readonly statusCode: number;
  public readonly code: string;

  constructor(message: string, statusCode: number = 500, code: string = 'INTERNAL_ERROR') {
    super(message);
    this.statusCode = statusCode;
    this.code = code;
    Object.setPrototypeOf(this, new.target.prototype);
  }
}

export class ValidationError extends ApplicationError {
  constructor(message: string) {
    super(message, 400, 'VALIDATION_ERROR');
  }
}

export class AuthorizationError extends ApplicationError {
  constructor(message: string = 'Insufficient permission scopes') {
    super(message, 403, 'AUTHORIZATION_ERROR');
  }
}

File: shared/utils/error_handlers.py
class ApplicationError(Exception):
    def __init__(self, message: str, status_code: int = 500, code: str = "INTERNAL_ERROR"):
        super().__init__(message)
        self.message = message
        self.status_code = status_code
        self.code = code

class ValidationError(ApplicationError):
    def __init__(self, message: str):
        super().__init__(message, status_code=400, code="VALIDATION_ERROR")

class AuthorizationError(ApplicationError):
    def __init__(self, message: str = "Insufficient permission scopes"):
        super().__init__(message, status_code=403, code="AUTHORIZATION_ERROR")
